File: tcp/set_connect_timeout/cxx/tcp_ping.cpp
#include <string>
#include <sstream>
#include <iostream>
#include <gflags/gflags.h>

#include "unp.hpp"

void connect_to(const std::string& host, const std::string& serv) {
    std::vector<SocketAddressInfo> infolist;
    try 
    {
        infolist = Getaddrinfo(
                host, serv, 0, SOCK_STREAM, 0,
                AI_ADDRCONFIG | AI_V4MAPPED | AI_CANONNAME
                );
    }
    catch (const std::system_error& e)
    {
        std::cout << "Name service failure: " << e.what() << std::endl;
        exit(1);
    }

    auto& info = infolist[0];
    Socket s(info.family, info.type, info.protocol);
    try 
    {
        s.Connect(info.address, 5);
        std::cout << "Success: host " << info.canon_name
                << " is listening on port " << info.address.IPPort()
                << std::endl;
    }
    catch (const std::system_error& e)
    {
        std::cout << "Network failure: " << e.what() << std::endl;
    }
}

std::string usage(const char* prog) {
    std::ostringstream os;
    os << "Try connecting by tcp\n\n"
        << "Usage: " << prog << " [--help] [--host HOST] [--serv SERV]";
    return os.str();
}

DEFINE_string(host, "", "hostname that you want to contact");
DEFINE_string(serv, "www", "TCP service name or port number");

int main(int argc, char *argv[]) {
    gflags::SetUsageMessage(usage(argv[0]));
    gflags::ParseCommandLineFlags(&argc, &argv, true);

    if (FLAGS_host.empty()) {
        std::cout << usage(argv[0]) << std::endl;
        exit(1);
    }

    connect_to(FLAGS_host, FLAGS_serv);

    return 0;
}

File: dns/tcp_ping/python/tcp_ping.py
#!/usr/bin/env python3
# Foundations of Python Network Programming, Third Edition
# https://github.com/brandon-rhodes/fopnp/blob/m/py3/chapter04/www_ping.py
# Find the WWW service of an arbitrary host using getaddrinfo().

import argparse, socket, sys

def connect_to(hostname_or_ip, servname_or_port):
    try:
        infolist = socket.getaddrinfo(
            hostname_or_ip, servname_or_port, 0, socket.SOCK_STREAM, 0,
            socket.AI_ADDRCONFIG | socket.AI_V4MAPPED | socket.AI_CANONNAME,
            )
    except socket.gaierror as e:
        print('Name service failure:', e.args[1])
        sys.exit(1)

    info = infolist[0]  # per standard recommendation, try the first one
    socket_args = info[0:3]
    address = info[4]
    s = socket.socket(*socket_args)
    try:
        s.connect(address)
    except socket.error as e:
        print('Network failure:', e.args[1])
    else:
        print('Success: host', info[3], 'is listening on port', address[1])

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Try connecting by tcp')
    parser.add_argument('--host', type=str, required=True, help='hostname that you want to contact')
    parser.add_argument('--serv', type=str, default='www', help='TCP service name or port number (default: %(default)s)')
    args = parser.parse_args()
    connect_to(args.host, args.serv)
